### PR TITLE
Add nix build tooling

### PR DESCRIPTION
**Pull request by robin-nitrokey**
This patch adds a nix-shell environment and a Makefile to build the firmware binaries using Nix.  The idea is to make it easier to build a previous version of the firmware the same way it was built at the time of development.  Potentially, it could also be used with the CI for reproducible builds (to be investigated).

### CODE REVIEW

File: utils/nix/Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+RUNNER ?= ../../runners/embedded
+RUNNER_NK3AM ?= $(RUNNER)/artifacts/runner-nrf52-bootloader-nk3am.bin
+RUNNER_NK3XN ?= $(RUNNER)/artifacts/runner-lpc55-nk3xn.bin
+
+.PHONY: all
+all:
+	@nix-shell --run "make binaries"
```
Comment (sosthene-nitrokey) [inline utils/nix/Makefile:7]: I don't really like that it does pass through any make flags passed to `make all`.

I would have put only the `shell.nix` file here as a way to install all dependencies. What is the benefit of having the makefile here compared to having the nix shell definition, starting a shell with it and using the other Makefiles?

Comment (robin-nitrokey) [reply utils/nix/Makefile:7]: We have changed the build process previously and we might do it again.  Having the Makefile allows us to just checkout a revision, run `make -C utils/nix` and be sure that it works.  If we have to manually create the nix shell and call the correct make targets, things might get complicated if we change the runner Makefile.  We could also move the target to the root Makefile if you prefer that.

Also, adding the `--pure` flag to the `nix-shell` invocation should make sure that nothing is passed through.

Comment (sosthene-nitrokey) [reply utils/nix/Makefile:7]: Ok, looks good

**Review by sosthene-nitrokey — APPROVED** [review]
I can't really test the shell.nix file but it looks good to me.

If you've tested it once with `--pure` so we're sure it's not missing anything it's ok for me to merge.

**Comment by robin-nitrokey** [comment]
Added the `--pure` flag to `nix-shell` and the `git` dependency to `shell.nix`.

**Pull request merged**